### DISCUSSION TITLE
Comparison methods

### DISCUSSION THREAD

**Discussion by wellington36** [wellington36/stan_summer#5] [General]
We create an equivalent version of [truncation_tests/adaptive_versus_threshold.r](https://github.com/maxbiostat/truncation_tests/blob/main/tests/adaptive_versus_threshold.r) (using the library sumR) for this repository using Stan [comparison_methods.R](https://github.com/wellington36/stan_summer/blob/comparison_methods/R/comparison_methods.R). Some of the differences between the outputs are presented here.

### L = 0.5 (adaptive_versus_threshold.r / comparison_methods.R)
![image](https://github.com/wellington36/stan_summer/assets/61877847/348ccef6-f660-4a79-ab70-c2c60bdd16ad) 
![image](https://github.com/wellington36/stan_summer/assets/61877847/8eff6231-a77a-49f7-a6fc-c729425ae1f9)

###  L = 0.909 (adaptive_versus_threshold.r / comparison_methods.R)
![image](https://github.com/wellington36/stan_summer/assets/61877847/8f61f366-4623-4864-9494-0c0b1878b48b)
![image](https://github.com/wellington36/stan_summer/assets/61877847/5d18e77b-73d5-4020-a2de-0f9125cfbfcf)

###  true_method = Huge (adaptive_versus_threshold.r / comparison_methods.R)
![image](https://github.com/wellington36/stan_summer/assets/61877847/2ffad683-2b47-4a78-9947-912b60e83211)
![image](https://github.com/wellington36/stan_summer/assets/61877847/c227a97b-11e3-487c-a9b8-148dbe28805c)

_I believe that the differences are being caused by the difference in origin of the basic functions (but I do not exclude an error in my implementation)._

*Edit: Adding parameter `sig_figs = 18` in comparison_methods.R for precision.*

@maxbiostat

**Comment by wellington36** [comment]
@maxbiostat The problem with the code is probably numerical inaccuracy, after a certain point, regardless of whether you add more terms or not, the value is the same.

Above in the `error` column we compared the Stan implementations with the logFunction in Cpp (with long double that provides 80-bit precision) evaluated in 1e6 terms which apparently has better accuracy than the Stan implementation (with real that provides 64-bit precision).

```R
lapply(ext[grep("estimated", names(ext))],
+        function(x) Rmpfr::mpfr(x[1], 1000))
$estimatedSumToThreshold
1 'mpfr' number of precision  1000   bits 
[1] -0.5408720000000000194262383956811390817165374755859375

$estimatedErrorBoundingPairs
1 'mpfr' number of precision  1000   bits 
[1] -0.5408720000000000194262383956811390817165374755859375

$estimatedBatches
1 'mpfr' number of precision  1000   bits 
[1] -0.5408720000000000194262383956811390817165374755859375

$estimatedBigSum
1 'mpfr' number of precision  1000   bits 
[1] -0.5408720000000000194262383956811390817165374755859375
```

What I can conclude is that the way we are implementing Stan, at least we have less precision than in cpp.

**Reply by maxbiostat** [reply]
I think this is sensible. The question is: can one change the Stan implementation to have better precision at little computational cost? Most people are more than happy to have $\varepsilon < 1E-10$. There is no point in providing an approximation that is correct up to the 20th decimal place. In the context of these approximations, all we need is to beat (Markov chain) Monte Carlo error by a few orders of magnitude.

**Reply by wellington36** [reply]
Of course, but the actual problem is that if I evaluate the firsts 42 terms of the following logFunction with the `sumR`, we have an error of 1e-16, but with the `SumToCap` we have an error of 1e-8 (this can also be seen in all methods, see the first comment in the discussion).

```stan
real logFunction(int k, array[] real p){
  return - 2*log(k+1) - (k+1)*log(p[1]);
}
```

**Reply by maxbiostat** [reply]
So let me see if I get this right: Stan is returning significantly different approximations for low numbers of iterations? Can you confirm this happens for different values of  `p`  for the series above?

**Reply by wellington36** [reply]
No, the significant difference is between the library sumR (in C) and the Stan. For exemplification:
#### Usind the sumR
```R
library(sumR)

Rcpp::cppFunction(code='
  double compute_lterm(long n, Rcpp::NumericVector p)
  {
  long double ans = - 2*log(n+1) - (n+1)* log(p[0]);
  return ans;
  }')

TTV.1 <- log(1/12 * (pi^2 - 6*log(2)^2)) # true value

naive.1 <- sumR::infiniteSum(
  logFunction = compute_lterm,
  parameters = 2,
  epsilon = .Machine$double.eps,
  maxIter = 1E6,
  n0 = 0,
  forceAlgorithm = 1
)

print(naive.1$n)
Rmpfr::mpfr(naive.1$sum - TTV.1, 1000)
```
*Out:*
```R
> print(naive.1$n)
[1] 41
> Rmpfr::mpfr(naive.1$sum - TTV.1, 1000)
1 'mpfr' number of precision  1000   bits 
[1] -1.1102230246251565404236316680908203125e-16
```
#### Now with our stan
*First, the SumToCap*
```stan
array[] real SumToCap(array[] real p, int nmax, int n0) {
  vector[nmax + 1] storeVal;
  int n = nmax + 1;
  for (i in 1:n){
    storeVal[i] = logFunction(n0 + i - 1, p);
  }
  return {log_sum_exp(sort_asc(storeVal[:n])), 1. * n};
}
```

*Second, we define the logFunction (that I called sandbox.stan)*
```stan
functions{
  real logFunction(int k, array[] real p){
    return - 2*log(k+1) - (k+1)*log(p[1]);
  }
  // logFunction must be defined beforehand
  #include SumToCap.stan
}

data {
  real<lower=0> p;
  int<lower=0> n;
  int<lower=0> n0;
}

transformed data {
  array[1] real params = {p};
}

generated quantities {
  array[2] real estimatedSum = SumToCap(params, n, n0);
}
```
*And I run*
```R
library(cmdstanr)

stanfit <- function(fit) rstan::read_stan_csv(fit$output_files())

model <- cmdstan_model("stan/sandbox.stan",
                       include_paths = "stan")

TTV.1 <- log(1/12 * (pi^2 - 6*log(2)^2))

cdata <- list(
  p = 2,
  n = 40,
  n0 = 0
)

comparison <- stanfit(
  model$sample(
    data = cdata,
    fixed_param = TRUE,
    iter_sampling = 1,
    adapt_engaged = FALSE,
    chains = 1
  )
)

( ext <- rstan::extract(comparison) )

print(ext$estimatedSum[2])
Rmpfr::mpfr(ext$estimatedSum[1] - TTV.1, 1000)

system("rm stan/sandbox")
```
*Out:*
```R
> print(ext$estimatedSum[2])
[1] 41
> Rmpfr::mpfr(ext$estimatedSum[1] - TTV.1, 1000)
1 'mpfr' number of precision  1000   bits 
[1] -3.5908749451696309051840216852724552154541015625e-7
```

Thats the problem, with the same number of terms our version is 8th decimals worse than the version in `sumR`, even doing just one sum. I don't see a reason for this difference.

**Reply by maxbiostat** [reply]
This is what I meant : Stan I'd giving different answers when compared to sumR. Is it possible the sumR answer is wrong somehow?

**Reply by wellington36** [reply]
I don't think so, given that it is in accordance with the true value.

**Reply by maxbiostat** [reply]
So something in the Stan implementation is leading to a catastrophic loss of precision. Can you please test a few values for the `p` parameter? I want to understand if the error difference is dependent on the "hardness" of the problem.

**Reply by wellington36** [reply]
Of course, some different tests (I fixed sumToCap to use the same amount of terms as sumR):

p / n  |  sumR   | sumToCap
---|----|---
1.000001 / 5138940 | 1.00e-10 | 2.95e-7
1.00001 / 869284 | 1.11e-11 | 2.30e-7
1.0001 / 125622 | 1.17e-12 | 4.21e-7 
1.001 / 16615 | 1.21e-13 | 3.27e-7
1.01 / 2086 | 1.28e-14 | 3.13e-7
1.1 / 261 | 1.55e-15 | 3.31e-8
1.5 / 68| 1.38e-16 | 2.95e-7
2 / 41| 1.11e-16 | 3.59e-7
3 / 26 | 4.44e-16 | 4.80e-7

**Reply by maxbiostat** [reply]
I think sumR is using error bounding pairs. The argument forceAlgorithm=1 uses the Braden-inspired method. A more accurate comparison would be with finiteSum, I think.

**Reply by wellington36** [reply]
No, forceAlgorithm=1 uses sumToThreshold. I think the stopping method in this case doesn't matter, since I wait for the algorithm to say how many terms were needed and I run the sumToCap with the exact number of terms.

**Reply by maxbiostat** [reply]
Right. Let's start debugging, then. Can you remove the sorting inside logsumexp in the Stan code and see if that makes any difference?

**Reply by wellington36** [reply]
For the case p=2, there are no difference.

**Reply by wellington36** [reply]
I think it's not really an error in the code since the same thing doesn't just happen with sumToCap, but with the other methods in Stan. For `p=2`, the first term is 0.0624 and the 41th is 1.2889e-16 for the R this is less than .Machine$double.eps, then is equal to zero.

**Reply by maxbiostat** [reply]
Alright. It's not clear to me how to proceed here. Do you wanna have a crack at reimplementing or fixing the other methods in Stan?

**Reply by wellington36** [reply]
Of course, I will try to make some headway until our meeting.

**Comment by wellington36** [comment]
I think that the problem is in the language. More specifically in the `log` function. If I define a stan code for simply evaluate the logarithm and test it, we can see some impression.
*sandbox.stan*
```stan
data {
  real x;
}

generated quantities {
  real defauldLog = log(x);
}
```
*sandbox.R*
```R
library(cmdstanr)

stanfit <- function(fit) rstan::read_stan_csv(fit$output_files())

model <- cmdstan_model("sandbox.stan",
                       include_paths = "stan")
x <- 2
cdata <- list(
  x = x
)

comparison <- stanfit(
  model$sample(
    data = cdata,
    fixed_param = TRUE,
    iter_sampling = 1,
    adapt_engaged = FALSE,
    chains = 1
  )
)

( ext <- rstan::extract(comparison) )

#here we check the difference between the evaluation in stan and in R
print("Ln 2 in R:")
Rmpfr::mpfr(log(x), 100)

print("Ln 2 in Stan")
Rmpfr::mpfr(ext$dlog, 100)

print("Difference:")
Rmpfr::mpfr(abs(ext$dlog - log(x)), 100)

system("rm sandbox")
```
*Out:* 1.8055994532861774359844275750220e-7

@maxbiostat  This means that the stan implementation differs from R in the seventh decimal place. But the R is right with 16th decimal places, this mean that the log in stan is inaccurate (I test with other values for the log all worst than 10^7).

**Reply by wellington36** [reply]
This inaccuracy is strange, but I don't see how it's not a problem with the library. Maybe we should open an issue commenting on this.

**Reply by wellington36** [reply]
For this case ($ln 2$), the value of R and of Stan:

```R
> Rmpfr::mpfr(log(2), 100)
1 'mpfr' number of precision  100   bits 
[1] 0.69314718055994528622676398299518
> 
> Rmpfr::mpfr(ext$defauldLog, 100)
'mpfrArray' of dim(.) =  (1) of precision  100   bits 
[1] 0.69314699999999995760902038455242
```
